File: 0018-4sum/0018-4sum.cpp
// 1 last try

class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());
        int siz = nums.size();
        for(int i = 0; i < siz; i++){
            for(int j = i + 1; j < siz; j++){
                int start = j + 1;
                int end = siz - 1;
                long long sum = (long long)target - (long long)nums[i] - (long long)nums[j];
                while(start < end){
                    if(nums[start] + nums[end] == sum){
                        ans.push_back({nums[i], nums[j], nums[start], nums[end]});
                        int third = nums[start];
                        int forth = nums[end];
                        while(start < end && nums[start] == third) start++;
                        while(start < end && nums[end] == forth) end--;
                    }
                    else if(nums[start] + nums[end] > sum) end--;
                    else start++;
                }
                while(j + 1 < siz && nums[j + 1] == nums[j]) j++;
            }
            while(i + 1 < siz && nums[i + 1] == nums[i]) i++;
        }
        return ans;
    }
};

// Time: O(n^3)
// Space: O(1)

File: 0018-4sum/0018-4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        res = []
        n = len(nums)
        for i in range(n - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            for j in range(i + 1, n - 2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                new_target = target - (nums[i] + nums[j])
                pairs = self.twoSum(nums, new_target, j + 1)
                for pair in pairs:
                    res.append([nums[i], nums[j]] + pair)
        return res
    
    def twoSum(self, nums, target, start):
        left, right = start, len(nums) - 1
        res = []
        while left < right:
            summation = nums[left] + nums[right]
            if target == summation:
                res.append([nums[left], nums[right]])
                third, fourth = nums[left], nums[right]
                while left < right and nums[left] == third:
                    left += 1
                while left < right and nums[right] == fourth:
                    right -= 1
            elif target > summation:
                left += 1
            else:
                right -= 1
        return res